src/spacy/crux.clj

```diff
@@ -42,19 +42,52 @@
 (defn- add-event-id [event-id doc]
   (assoc doc ::belongs-to-event event-id))
 
-(defn- persist! [node {::domain/keys [facts event]}]
+(defn- replace!
+  "Stores the given event and facts iff the current event state in the database
+  is equal to old-state. This allows us to make sure there were no concurrent
+  writes."
+  [node old-state {::domain/keys [facts event]}]
   (assert (s/valid? ::domain/event event)
           (s/explain-str ::domain/event event))
   (assert (s/valid? ::domain/facts facts)
           (s/explain-str ::domain/facts facts))
   (let [event-id (:crux.db/id event)
         new-facts (->> facts
                        (map (partial add-event-id event-id))
-                       (map maybe-add-crux-id))]
-    (crux/await-tx
-     node
-     (crux/submit-tx node (for [doc (cons event new-facts)]
-                            [:crux.tx/put doc])))))
+                       (map maybe-add-crux-id))
+        tx (crux/submit-tx node (cons
+                                 [:crux.tx/match event-id old-state]
+                                 (for [doc (cons event new-facts)]
+                                   [:crux.tx/put doc])))]
+    (crux/await-tx node tx)
+    (if-not (crux/tx-committed? node tx)
+      {::error ::concurrent-writes})))
+
+(def ^:private retries 5)
```

Comment:
```suggestion
(def ^:private retries 10)
```

I've now tested this locally, and with Postgres running instead of SQLite, and it definitely works a lot better. I think I was getting most of my errors from the SQLite database under the hood, and not actually from this code.

However, I still get a few 400 responses when running the test locally (maybe 2 or 3 each time) with 5 retries. When I bumped it up to 10 retries, I get pretty much no 400 responses at all when running the script: I got an occasional single 400 response a few times when I ran the test on an empty database, and when the number of sessions increases drastically (> 300), the number of 400 responses became slightly more frequent.

My suggestion would be that we increase the retries to 10 just to be on the safe side. That would theoretically mean that the response from spacy could theoretically be after 6.5 seconds (`(reduce + (for [x (range 0 10)] (* 100 (+ 2 x)))) => 6500`), but that would really only be an upper limit, and I don't expect that in the real world we would have the same intensity of concurrent requests that we have when running the script.

Reply:
Sounds great, thanks for experimenting with it!